gusto/linear_solvers.py

```diff
@@ -389,72 +393,77 @@ def V(u):
         _dl = TestFunction(Vtrace)
-        tM = assemble(_dl('+')*_l0('+')*(dS_v + dS_h)
-                      + _dl*_l0*ds_v + _dl*_l0*(ds_t + ds_b))
+        a_tr = _dl('+')*_l0('+')*(dS_vp + dS_hp) + _dl*_l0*ds_vp + _dl*_l0*ds_tbp
 
-        Lrhobar = Function(Vtrace)
-        Lpibar = Function(Vtrace)
-        rhopi_solver = LinearSolver(tM, solver_parameters={'ksp_type': 'cg',
-                                                           'pc_type': 'bjacobi',
-                                                           'sub_pc_type': 'ilu'},
-                                    options_prefix='rhobarpibar_solver')
+        def L_tr(f):
+            return _dl('+')*avg(f)*(dS_vp + dS_hp) + _dl*f*ds_vp + _dl*f*ds_tbp
+
+        cg_ilu_parameters = {'ksp_type': 'cg',
+                             'pc_type': 'bjacobi',
+                             'sub_pc_type': 'ilu'}
 
         # Project field averages into functions on the trace space
         rhobar_avg = Function(Vtrace)
         pibar_avg = Function(Vtrace)
 
-        def _traceRHS(f):
-            return (_dl('+')*avg(f)*(dS_v + dS_h)
-                    + _dl*f*ds_v + _dl*f*(ds_t + ds_b))
+        rho_avg_prb = LinearVariationalProblem(a_tr, L_tr(rhobar), rhobar_avg)
+        pi_avg_prb = LinearVariationalProblem(a_tr, L_tr(pibar), pibar_avg)
 
-        assemble(_traceRHS(rhobar), tensor=Lrhobar)
-        assemble(_traceRHS(pibar), tensor=Lpibar)
+        rho_avg_solver = LinearVariationalSolver(rho_avg_prb,
+                                                 solver_parameters=cg_ilu_parameters,
+                                                 options_prefix='rhobar_avg_solver')
+        pi_avg_solver = LinearVariationalSolver(pi_avg_prb,
+                                                solver_parameters=cg_ilu_parameters,
+                                                options_prefix='pibar_avg_solver')
 
-        # Project averages of coefficients into the trace space
         with timed_region("Gusto:HybridProjectRhobar"):
-            rhopi_solver.solve(rhobar_avg, Lrhobar)
+            rho_avg_solver.solve()
 
         with timed_region("Gusto:HybridProjectPibar"):
-            rhopi_solver.solve(pibar_avg, Lpibar)
-
-        # Add effect of density of water upon theta
-        if self.moisture is not None:
-            water_t = Function(Vtheta).assign(0.0)
-            for water in self.moisture:
-                water_t += self.state.fields(water)
-            theta_w = theta / (1 + water_t)
-            thetabar_w = thetabar / (1 + water_t)
-        else:
-            theta_w = theta
-            thetabar_w = thetabar
+            pi_avg_solver.solve()
 
         # "broken" u, rho, and trace system
-        eqn = (inner(w, (state.h_project(u) - u_in))*dx
-               - beta_cp*div(theta_w*V(w))*pibar*dxp
-               # following does nothing but is preserved in the comments
-               # to remind us why (because V(w) is purely vertical).
```

Comment:
I notice we've lost this comment and the commented out term is now added - do we need to add it if it does nothing or is the comment no longer true?

Reply:
Okay I'm reasonably convinced it's not needed. I've gone ahead and updated the branch